File: public/docs/_examples/toh-6/ts/app/hero.service.ts
// #docplaster
// #docregion , imports
import { Injectable }    from '@angular/core';
import { Headers, Http } from '@angular/http';

// #docregion rxjs
import 'rxjs/add/operator/toPromise';
// #enddocregion rxjs

import { Hero } from './hero';
// #enddocregion imports

@Injectable()
export class HeroService {

  // #docregion update
  private headers = new Headers({'Content-Type': 'application/json'});
  // #enddocregion update
  // #docregion getHeroes
  private heroesUrl = 'app/heroes';  // URL to web api

  constructor(private http: Http) { }

  getHeroes(): Promise<Hero[]> {
    return this.http.get(this.heroesUrl)
    // #docregion to-promise
               .toPromise()
    // #enddocregion to-promise
    // #docregion to-data
               .then(response => response.json().data as Hero[])
    // #enddocregion to-data
    // #docregion catch
               .catch(this.handleError);
    // #enddocregion catch
  }
  // #enddocregion getHeroes

  getHero(id: number): Promise<Hero> {
    return this.getHeroes()
               .then(heroes => heroes.find(hero => hero.id === id));
  }

  // #docregion delete
  delete(id: number): Promise<void> {
    let url = `${this.heroesUrl}/${id}`;
    return this.http.delete(url, {headers: this.headers})
      .toPromise()
      .then(() => null)
      .catch(this.handleError);
  }
  // #enddocregion delete

  // #docregion create
  create(name: string): Promise<Hero> {
    return this.http
      .post(this.heroesUrl, JSON.stringify({name: name}), {headers: this.headers})
      .toPromise()
      .then(res => res.json().data)
      .catch(this.handleError);
  }
  // #enddocregion create
  // #docregion update

  update(hero: Hero): Promise<Hero> {
    const url = `${this.heroesUrl}/${hero.id}`;
    return this.http
      .put(url, JSON.stringify(hero), {headers: this.headers})
      .toPromise()
      .then(() => hero)
      .catch(this.handleError);
  }
  // #enddocregion put, update

  // #docregion handleError
  private handleError(error: any): Promise<any> {
    console.error('An error occurred', error); // for demo purposes only
    return Promise.reject(error.message || error);
  }
  // #enddocregion handleError
}

File: public/docs/_examples/toh-6/dart/lib/hero_service.dart
// #docplaster
// #docregion , imports
import 'dart:async';
import 'dart:convert';

import 'package:angular2/core.dart';
import 'package:http/http.dart';

import 'hero.dart';
// #enddocregion imports

@Injectable()
class HeroService {
  // #docregion update
  static final _headers = {'Content-Type': 'application/json'};
  // #enddocregion update
  // #docregion getHeroes
  static const _heroesUrl = 'app/heroes'; // URL to web API

  final Client _http;

  HeroService(this._http);

  Future<List<Hero>> getHeroes() async {
    try {
      final response = await _http.get(_heroesUrl);
      final heroes = _extractData(response)
          .map((value) => new Hero.fromJson(value))
          .toList();
      return heroes;
      // #docregion catch
    } catch (e) {
      throw _handleError(e);
    }
    // #enddocregion catch
  }

  // #docregion extract-data
  dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];
  // #enddocregion extract-data

  // #docregion handleError
  Exception _handleError(dynamic e) {
    print(e); // for demo purposes only
    return new Exception('Server error; cause: $e');
  }
  // #enddocregion handleError, getHeroes

  Future<Hero> getHero(int id) async =>
      (await getHeroes()).firstWhere((hero) => hero.id == id);

  // #docregion create
  Future<Hero> create(String name) async {
    try {
      final response = await _http.post(_heroesUrl,
          headers: _headers, body: JSON.encode({'name': name}));
      return new Hero.fromJson(_extractData(response));
    } catch (e) {
      throw _handleError(e);
    }
  }
  // #enddocregion create
  // #docregion update

  Future<Hero> update(Hero hero) async {
    try {
      var url = '$_heroesUrl/${hero.id}';
      final response =
          await _http.put(url, headers: _headers, body: JSON.encode(hero));
      return new Hero.fromJson(_extractData(response));
    } catch (e) {
      throw _handleError(e);
    }
  }
  // #enddocregion update

  // #docregion delete
  Future<Null> delete(int id) async {
    try {
      var url = '$_heroesUrl/$id';
      await _http.delete(url, headers: _headers);
    } catch (e) {
      throw _handleError(e);
    }
  }
  // #enddocregion delete

}
